<html lang="en" data-theme="dark"><head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Debugging Debrief #2: Deep diving into the thread pool</title>
    <link rel="stylesheet" href="https://unpkg.com/@picocss/pico@latest/css/pico.min.css">
    <style>
      h1 {
        margin-top: 20px;
      }

      .back-btn {
        position: absolute;
        left: 20;
        top: 20;
      }

      blockquote {
        padding: 0px 10px;
      }

      code {
        color: #a5cce6;
      }

      pre code {
        color: #bfd4e2;
      }

      .container {
        max-width: 800px;
      }

      p {
        font-size: 18px;
      }

      img {
        max-width: 400px;
        text-align: center;
        margin: auto;
        display: block;
      }

      .column {
        float: left;
        padding: 5px;
      }

      .img-gallery-2 .column {
        width: 50%;
      }
      
      .row::after {
        content: "";
        clear: both;
        display: table;
      }
    </style>

    <script>
      !(function (t, e) {
        var o, n, p, r;
        e.__SV ||
          ((window.posthog = e),
          (e._i = []),
          (e.init = function (i, s, a) {
            function g(t, e) {
              var o = e.split(".");
              2 == o.length && ((t = t[o[0]]), (e = o[1])),
                (t[e] = function () {
                  t.push([e].concat(Array.prototype.slice.call(arguments, 0)));
                });
            }
            ((p = t.createElement("script")).type = "text/javascript"),
              (p.async = !0),
              (p.src = s.api_host + "/static/array.js"),
              (r = t.getElementsByTagName("script")[0]).parentNode.insertBefore(
                p,
                r
              );
            var u = e;
            for (
              void 0 !== a ? (u = e[a] = []) : (a = "posthog"),
                u.people = u.people || [],
                u.toString = function (t) {
                  var e = "posthog";
                  return (
                    "posthog" !== a && (e += "." + a), t || (e += " (stub)"), e
                  );
                },
                u.people.toString = function () {
                  return u.toString(1) + ".people (stub)";
                },
                o =
                  "capture identify alias people.set people.set_once set_config register register_once unregister opt_out_capturing has_opted_out_capturing opt_in_capturing reset isFeatureEnabled onFeatureFlags".split(
                    " "
                  ),
                n = 0;
              n < o.length;
              n++
            )
              g(u, o[n]);
            e._i.push([i, s, a]);
          }),
          (e.__SV = 1));
      })(document, window.posthog || []);
      posthog.init("Tom1k0Lh0sRN7EK8LIlU5mbbmAHMBSk37lc02ZQ__38", {
        api_host: "https://app.posthog.com",
      });
    </script>
  </head>
  <body>
    <div class="back-btn">
      <a href="/">‚Üê Home</a>
    </div>
    <main class="container" id="page-content"><h1 id="debugging-debrief-2-deep-diving-into-the-thread-pool">Debugging Debrief #2: Deep diving into the thread pool</h1>
<blockquote>
<p><small>This piece is a part of my <em>Debugging Debrief</em> series, which includes writeups of tracking down and fixing bugs. My goal is that this series motivates me to write about software more often, so I can kick back into the habit, reinforce my learnings through the writing, and refer back to these posts in the future.</small></p>
</blockquote>
<p>Some months ago I found myself with the following issue in my sprint priorities:</p>
<p><a href="https://github.com/PostHog/bigquery-plugin/issues/10">"BigQuery is very flakey on local / self-hosted"</a></p>
<p>Keeping context to a minimum, we run an open source analytics platform that you can self-host, and there are plugins to help you send data out of the platform and into other tools, of which BigQuery is one of them.</p>
<p>Now, two weeks or so before the sprint at hand, a customer who was using the BigQuery plugin on their self-hosted instance reported seeing a lot of errors on the plugin's logs, saying something about execution having timed out.</p>
<p>Plugins can run tasks that last up to 30 seconds at a time, and it seemed inserts to BigQuery were taking more than that to complete. </p>
<p>Even worse, it seemed that somehow some inserts were succeeding despite throwing the error, which caused us to export duplicates as our retry logic was triggered.</p>
<p>Plus, to top it off, this was reportedly happening on local Postgres-backed deployments and self-hosted ClickHouse-backed instances, but not on our ClickHouse-backed Cloud instance. What?</p>
<p>And so, back to my sprint goals. Having read all the context I could find, it seemed the consensus until now was that BigQuery was at fault here. </p>
<p>However, I was a bit skeptical. I had been working on building plugins, and just the week before I had encountered similar cryptic errors while working on another plugin.</p>
<p>I originally shrugged them off, but the light bulb went on as soon as I read the issue about the BigQuery plugin. </p>
<p>Something deeper was going on.</p>
<h2 id="the-util-shuffle">The util shuffle</h2>
<p>While the BigQuery plugin didn't seem to be the only one affected, it could potentially give me some insight into what's happening. </p>
<p>I started running it while inspecting the code, and something quickly became clear: we'd regularly throw errors on the first insert attempt, but never on retries.</p>
<p>What could be different here?</p>
<p>Bingo! The buffer.</p>
<p>Our plugins have a few options when it comes to asynchronous processing, primarily the jobs API and the buffer util.</p>
<p>The BigQuery plugin used the buffer on the first insert attempt, but leveraged jobs (another API) to process retries.</p>
<p>Given that the problem seemed to only occur on the first run, it seemed that the buffer was to blame. </p>
<h2 id="beyond-a-reasonable-doubt">Beyond a reasonable doubt</h2>
<p>Ok, so the buffer seemed to be the problem, but this needed to be confirmed.</p>
<p>As such, I setup a plugin that used both a buffer and a job to call the same async function that made a request to an endpoint that'd always wait 5s before responding, and timed the performance of the two.</p>
<p>The job always executed in around the same amount of time, but the buffer would vary a lot:</p>
<p><img src="https://user-images.githubusercontent.com/38760734/123464398-1efa6f00-d5c3-11eb-86a4-612fd36d0d3a.png" alt="logs"></p>
<p>I went through thousands of runs, and the job was always consistent. The buffer would vary from 5s to 29s and then came the timeouts.</p>
<p>So the buffer was the problem. But finding out why was where my beard hairs started to come out.</p>
<p>I started to notice connection reset, socket hang up, and other sorts of errors. I knew the request would only take 5s, so there were 25s to spare. This made no sense, and so my hunch was that this a problem lower down.</p>
<h2 id="üî•">üî•</h2>
<p>At this point I spent hours looking a Node.js and Piscina (our thread pool) docs and testing out different config options, but I wasn't going anywhere.</p>
<p>So finally I decided to run the plugin server with a flamegraph tool called <a href="https://github.com/davidmarkclements/0x">0x</a>.</p>
<p>And <em>that</em> looked like <em>this</em>:</p>
<p><img src="https://user-images.githubusercontent.com/38760734/123465219-3a19ae80-d5c4-11eb-857c-588c31a33262.png" alt="flamegraph"></p>
<p>Hm, what's <code>atomicsWaitLoop</code>?</p>
<p>I went to the Piscina repo and there was the answer:</p>
<pre><code>function atomicsWaitLoop (port : MessagePort, sharedBuffer : Int32Array) {
  if (!useAtomics) return;

  // This function is entered either after receiving the startup message, or
  // when we are done with a task. In those situations, the *only* thing we
  // expect to happen next is a 'message' on `port`.
  // That call would come with the overhead of a C++ ‚Üí JS boundary crossing,
  // including async tracking. So, instead, if there is no task currently
  // running, we wait for a signal from the parent thread using Atomics.wait(),
  // and read the message from the port instead of generating an event,
  // in order to avoid that overhead.
  // The one catch is that this stops asynchronous operations that are still
  // running from proceeding. Generally, tasks should not spawn asynchronous
  // operations without waiting for them to finish, though.
  
  ...
}
</code></pre>
<p>Oh.</p>
<blockquote>
<p>"Generally, tasks should not spawn asynchronous operations without waiting for them to finish, though."</p>
</blockquote>
<p>That's exactly what the buffer does!</p>
<p>So we were doing something our thread pool recommended against. That probably was it. But we did this across the board! How come this error didn't happen on Cloud but happened on self-hosted instances?</p>
<p>As it turns out, the error <em>did</em> happen on Cloud, but very infrequently. It was also merged in with other legitimate timeouts on Sentry, which made it hard to spot.</p>
<p>But that still wasn't it. What is <em>really</em> happening and how can we fix it? </p>
<h2 id="never-block-the-event-loop">Never block the event loop</h2>
<p>The golden rule in the JavaScript ecosystem is <a href="https://nodejs.org/en/docs/guides/dont-block-the-event-loop/">"don't block the event loop"</a>.</p>
<p>The event loop is JavaScript's approach to asynchronous operations in a single-threaded environment. </p>
<p>It is much better explained by <a href="https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/">others</a>, so I'll refrain from covering it here.</p>
<p>As it turns out, in order to improve performance around the communication between the main thread and the worker threads, Piscina defaults to running a (potentially infinite) while loop that proactively looks for new tasks for the worker on a specified port, rather than waiting for the callback triggered by listening for new messages.</p>
<p>As a result, while Piscina is checking for new messages, it blocks the event loop, so nothing else gets done. </p>
<p>And, given we have set timeouts for our tasks to execute, if a worker is blocking for more than 30 seconds, once it is free again for other processing, it goes to handle the callbacks from our task, but at this point we force the timeout. </p>
<p>Now, this only happens with <em>voided promises</em>. Promises that are explicitly voided (i.e. not awaited) in order to be processed "in the background", where the result is not important. If a promise is awaited, the task is not marked as complete as long as the promise is not done, and thus the worker doesn't get into the blocked state to look for new tasks.</p>
<p>Concretely, here's an example scenario:</p>
<ol>
<li>PostHog receives an event</li>
<li>This event is passed to a plugin for processing</li>
<li>This plugin voids a promise sending a POST request to BigQuery with the event</li>
<li>The worker marks the task as complete and goes on to look for new tasks</li>
<li>The event loop is blocked for 45 seconds</li>
<li>A new event comes in, and a new task is picked up</li>
<li>When it gets a chance, the event loop tries to handle the callback from the voided promise</li>
<li>As it goes to handle the callback from the request, our wrapper forces a timeout and the plugin errors (despite the request having completed successfully)</li>
<li>Our retry logic was triggered and the event was exported twice</li>
</ol>
<h2 id="when-the-lack-of-load-is-the-problem">When the lack of load is the problem</h2>
<p>Now, the reason this affected our clients' instances more than our Cloud instance is because the Piscina worker will only block the event loop when it is looking for new tasks. If it is processing a task, it will be able to handle the background callbacks and such normally.</p>
<p>So, while on Cloud we process hundreds of thousands of events per minute, with each event potentially triggering multiple tasks, some of our clients could feasibly go 30 seconds without a worker receiving a task, thus triggering this issue.</p>
<p>This is tricky, because upon seeing something like a timeout error, you might think you should scale <em>up</em>, as is the solution for other similar issues we have faced that cause timeouts. </p>
<p>However, if you scale vertically, we'll add more threads to the plugin server - the default setting is <code>n(threads) == n(cpus)</code> - and if you scale horizontally, we'll also be running more worker threads. </p>
<p>As such, you're increasing the chance that a given worker will have to wait 30 or more seconds for a task. </p>
<p>The (temporary) answer here would actually be to scale <em>down</em>.</p>
<h2 id="the-solution">The solution</h2>
<p>So, the client could scale down to potentially mitigate this problem, but that's certainly not a good solution.</p>
<p>But how did we <em>fix</em> this?</p>
<p>Well, one thing we could have done was get rid of voided promises altogether. However, the benefit they bring us in performance is significant.</p>
<p>As such, we started to brainstorm on approaches that would allow us to keep the voided promises.</p>
<p>The first attempt was very straightforward. Piscina provides a <code>useAtomics</code> flag that can be used to disable the blocking mechanism instead of the slower approach of waiting for the signal denoting a new message. </p>
<p>Upon benchmarking this with our system, however, I found that our event ingestion rate <a href="https://github.com/PostHog/plugin-server/issues/487#issuecomment-870509629">slowed down by around 27%</a> when <code>useAtomics</code> was set to <code>false</code>. This is an unacceptable performance drop, so we scrapped the idea.</p>
<p>Thus, after toying with a few other ideas, we landed on the following (note that we were already running a <em>fork</em> of Piscina):</p>
<p><strong>Introduce a setting denoting a maximum amount of time a worker should block the event loop while looking for tasks before falling back into the slower mechanism, defaulting it to 5 seconds. (<a href="https://github.com/PostHog/piscina/pull/4">Implementation</a>)</strong> </p>
<p>The reason this works is because on our Cloud instance, where performance is paramount, most of the time workers will get a new task in fractions of a second, so the absolute vast majority of tasks will be picked up using the fast mechanism.</p>
<p>However, on self-hosted instances processing low volumes, if workers are already taking e.g. 5 seconds to get a new task, the load on the instance is so little that the performance hit won't be problematic, since it is unlikely it will lead to a long backlog of tasks.</p>
<p>And so, that's how we kept our performance intact without touching the voided promises.</p>
<hr>
<p><em><a href="https://github.com/PostHog/plugin-server/issues/487">Here's the original ticket</a> where this was explained and a discussion about possibilities took place</em>.</p>
</main>
  

</body></html>